### ISSUE TITLE
Showing Header Menu on "contextmenu" event

### ISSUE THREAD

**Issue by smorandi** [ceolter/ag-grid#808]
I am currently evaluating different grids and ag-grid looks really great so far. However, i was wondering what the best possible way would be to trigger showing the header menu in case of a context menu event on the header row.

Currently i am doing it through a custom header cell renderer, i.e. "headerCellTemplate" where i set an additional eventlistener ("contextmenu" listener) on the eCell and then just triggering click() on the #agMenu. However that is quite hacky, hence i wonder if you could give me some hints there.

(As far as i've seen the RendererdHeaderCell just sets an event-listener for "click" on the ag-menu element. would it not be feasible to provide one for the context-menu event as well?)

**Comment by ceolter** [comment]
all of what you say makes sense in that a) it's the best way to do it given the options and b) it's not a great thing to have to do. however it will work fine.

i went to the code to see if i could add a 'show menu for column' api method quickly, but it's a bit tricky as the menu depends on the button for positioning, so i'd have to rethink that if to allow showing the menu without the button.

when you say context menu, are you talking about clicking anywhere on the header or just the menu button?

**Comment by smorandi** [comment]
yes, it should be activated by clicking anywhere on the according header cell, not just the button. i basically just need the button in my solution "as a relay" to the showMenu() method:

That's what i am basically doing atm:

```

this.gridOptions = <GridOptions>{
            getHeaderCellTemplate: params => this.getHeaderCellTemplate(params),
        };

private getHeaderCellTemplate(params:any) {

        var template:string = '<div class="ag-header-cell">' +
            '  <div id="agResizeBar" class="ag-header-cell-resize"></div>' +
            '  <span id="agMenu" class="ag-header-icon ag-header-cell-menu-button"></span>' +
            '  <div id="agHeaderCellLabel" class="ag-header-cell-label">' +
            '    <span id="agSortAsc" class="ag-header-icon ag-sort-ascending-icon"></span>' +
            '    <span id="agSortDesc" class="ag-header-icon ag-sort-descending-icon"></span>' +
            '    <span id="agNoSort" class="ag-header-icon ag-sort-none-icon"></span>' +
            '    <span id="agFilter" class="ag-header-icon ag-filter-icon"></span>' +
            '    <span id="agText" class="ag-header-cell-text"></span>' +
            '  </div>' +
            '</div>';

        var eTemplate:HTMLElement = Utils.loadTemplate(template);
        var column = params.column;
        var svgFactory = SvgFactory.getInstance();
        var gridOptionsWrapper = column.gridOptionsWrapper;

        this.addInIcon(eTemplate, gridOptionsWrapper, 'sortAscending', '#agSortAsc', column, svgFactory.createArrowUpSvg);
        this.addInIcon(eTemplate, gridOptionsWrapper, 'sortDescending', '#agSortDesc', column, svgFactory.createArrowDownSvg);
        this.addInIcon(eTemplate, gridOptionsWrapper, 'sortUnSort', '#agNoSort', column, svgFactory.createArrowUpDownSvg);
        this.addInIcon(eTemplate, gridOptionsWrapper, 'menu', '#agMenu', column, svgFactory.createMenuSvg);
        this.addInIcon(eTemplate, gridOptionsWrapper, 'filter', '#agFilter', column, svgFactory.createFilterSvg);

        var eMenu:any = eTemplate.querySelector('#agMenu');
        eTemplate.addEventListener("contextmenu", function (ev:Event) {
            ev.stopPropagation();
            ev.preventDefault();
            eMenu.click();
        });

        return eTemplate;
    }
```
It's a bit unfortunate, because i don't actually want anything to change in the header, so i am copying the boilerplate there and do the same stuff as the default is doing. i only would like to have the menu being shown an a context-menu event. just like in the "normal" cells. also it would be great to have the popup then positioned relative to the location of the click - and not actually where the agMenu-button is.
So, basically, the framework is already handling the context-menu event for the content cells, why not have the same handling for the header cells as well? from a UX point of view, IMHO, it would make a lot of sense to have the same "interaction" style may it be for normal or header cells. don't you agree?

**Comment by ceolter** [comment]
```don't you agree?``` -> maybe, 50% agree, yes for context menu, but what you want is the header normal menu to also be the header context. that's a bit weird.

**Comment by smorandi** [comment]
no, not really. what i would like to achieve ist just to have the possibility to activate the header menu by a  contextmenu event - nothing more. then you can still chose to either show the button or not. in our apps we try to strive for consistency in interaction styles. we support primarily context menu interactions. (our users like that - or just got used to it :) though for the sake of accesibility  it is always good to also provide alterative means for a command, so providing a button or keyboard shortcut is always beneficial...but that's just as a sidenote.

**Comment by ceolter** [comment]
```btw showing/hiding a column via the menu would also be great``` -> you can already doing this? assuming you are using ag-Grid enterprise here, one of the tabs has all the items.

**Comment by ceolter** [comment]
ok i looked at the code again. i think the best is for me to provide an API to show the menu, however the code would need to be change a bit as there is login in there that assumes the event is coming from a button, and it then positions the menu over that button.

i presume you need the menu to appear exactly where the click happens, ie mouse on top let of appearing menu?

**Comment by smorandi** [comment]
yes you are totally right, sorry! yeah, i think the best is to provide a little API. and yes, that's correct, the menu should appear relative to the position of the mouse click (x/y), just as for the cell context menu. would be really great to have this feature! thx a ton.

**Comment by ceolter** [comment]
ok done. i've implemented two new api methods:
```showColumnMenuAfterButtonClick(colKey, buttonElement)
showColumnMenuAfterMouseClick(colKey, mouseEvent)```

you will have to wait for the next release (at least two weeks away, maybe more), but i don't think this a blocker for you, it's just a usability issue. ok, now you can place your enterprise order!!!! :)

**Comment by smorandi** [comment]
If it was up to me...i would place a million orders...as it turns out i have to convince our "managers" that my POC using ag-grid is actually what they like too ;) 

(working in the financial industry i guess you know what that means...unless your managers are different than ours, hehe)

**Comment by ceolter** [comment]
yes, it means i have to be patient, and then you order one million developer seats. take your time. i'll start browsing some property websites.

**Comment by yarden-dvir** [comment]
I know this is an old thread but I have a question. I used the above showColumnMenuAfterMouseClick to get our header menu to show up when you right click the header. However, when I use this it shows the default ag grid header menu and not the one that we created. Any way around this?
